### ISSUE TITLE
General suitability for server side

### ISSUE THREAD

**Issue by pgillett** [ant-design-blazor/ant-design-blazor#1104]
### Issue

Is the library suitable for server side?

### Findings

We're currently implementing the library in a server side application. This has highlighted the sheer amount of messaging back and forth between the client and server.

We're having to rewrite sections of components so that we either offload the browser events to JS code, or simply omit things that we're not relying on.

We've got a test environment that simulates 30ms of latency (eg. equivalent of hosted off-site, access from home etc.), to check how responsive the UI is, and also a 500ms latency to check that component events don't clash. I appreciate 500ms is unrealistic, but it demonstrates where code appears to work, but is actually only working because it's keeping up faster than you can trigger new events.

### Sample data

This is the server network usage while moving the mouse around a date picker on the client (middle of the graph). It's hitting 1.5+ Mbps of data streaming (easy to get 2-3 Mbps by moving fast). Ok, not massive in data terms, but excessive for selecting a date. The beauty of Blazor SignalR is that it's occasional messages in bytes (or few kb max).

![image](https://user-images.githubusercontent.com/60133490/107040986-d7ac4700-67b7-11eb-9b4b-99dc49baa1e0.png)

Same situation server CPU - one client is adding nearly 20% usage. Presumably due to the constant rebuilding of the render tree and diffing.

![image](https://user-images.githubusercontent.com/60133490/107041136-09251280-67b8-11eb-9f37-974c60c42207.png)

### Thoughts

I appreciate that much of this is that the event implementation is in C#, and can't simply be offloaded to JS.

Would it be useful to highlight which controls work well under server side?

Or perhaps a global switch that simplifies the implementation? My previous issue #1022 highlighted this - eg. don't hook onkeypress, oninput etc.

**Comment by anddrzejb** [comment]
I agree with you that there are components that definitely need optimization. `DatePicker` is one of the biggest offenders - I noticed that some time ago but did not manage to get around to fixing it. Or at least try - this is a very involved task. I had a similar situation with menu component, when single menu change was calling `OnAfterRendering` more than 20 times. I managed to get that down to 2 times. `DatePicker` here is really baaaaad... Mind that menu component is a piece of cake compared to datepicker... I also tried to do similar work with a bit easier (but still complex) `Select` component but failed there... Mostly because I run of time. Anyways, I imagine a lot of components that require a lot of build-up are going to have this issue.

**Comment by ElderJames** [comment]
Thank you for the test @pgillett 

We are also currently working on covering Ant Design's features and fixing obvious bugs. And we have not yet optimised the resource consumption on the server side. But we would love to get any help in this part.

If you and your team has any ideas, just feel free to contact us and collaborate.

**Comment by pgillett** [comment]
@anddrzejb We've stripped Select down to a simple select where you add the options. We'll then create components locally to populate for specific data.

It uses JS to open/close and highlight. Doesn't implement keyboard yet.

**Comment by pgillett** [comment]
@ElderJames I'll have a look and see what we could do. We're duplicating code into new components just to strip bits out, so it'd be good to work from the same code
